// cucumber-html-report-generator/dist/test/e2e/pages/report-generation/report-generation-api-page.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReportGenerationApiPage = void 0;
const generateReport = __importStar(require("../../../../src/lib/generate-report/generate-report"));
const tsyringe_1 = require("tsyringe");
const path_1 = __importDefault(require("path"));
let ReportGenerationApiPage = class ReportGenerationApiPage {
    async generateReport(theme, folder) {
        const parameters = {
            openReportInBrowser: false,
            reportPath: folder,
            showExecutionTime: true,
            customStyle: `${path_1.default.resolve('./')}/src/resources/templates/css/style-dark-theme.css`,
            useCDN: false,
            featuresFolder: `${path_1.default.resolve('./')}/test/unit/data/features/correct/`,
            jsonDir: `${path_1.default.resolve('./')}/test/unit/data/cucumber-report-jsons/`,
            saveCollectedJSON: true,
            saveEnrichedJSON: true,
            saveReportInMongoDb: false,
            theme,
            reportMetadataTitle: 'Additional Data',
            reportMetadata: [
                { name: 'Project', value: 'custom project' },
                { name: 'Release', value: '1.2.3' },
                { name: 'Cycle', value: '2' },
                { name: 'Execution Start Date', value: '2021-03-26 14:01' },
                { name: 'Execution End Date', value: '2021-03-26 16:05' }
            ]
        };
        const reportId = await generateReport.generate(parameters);
        return reportId;
    }
};
ReportGenerationApiPage = __decorate([
    (0, tsyringe_1.injectable)()
], ReportGenerationApiPage);
exports.ReportGenerationApiPage = ReportGenerationApiPage;
