// cucumber-html-report-generator/dist/src/lib/models/report/results.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.reportResultsInitializer = exports.featureResultsInitializer = exports.scenarioResultsInitializer = exports.overviewInitializer = exports.beforeOrAfterOverviewInitializer = exports.moduleResultsInitializer = exports.featureModuleResultsInitializer = void 0;
const lodash = __importStar(require("lodash"));
const status_1 = require("./status");
const featureModuleResultsInitializer = () => lodash.cloneDeep({
    ambiguous: 0,
    ambiguousPercentage: '',
    failed: 0,
    failedPercentage: '',
    passed: 0,
    passedPercentage: '',
    pending: 0,
    pendingPercentage: '',
    skipped: 0,
    skippedPercentage: '',
    status: status_1.Status.passed,
    time: '',
    timeInNanoSeconds: 0,
    total: 0,
    undefined: 0,
    undefinedPercentage: '',
    various: 0,
    variousPercentage: ''
});
exports.featureModuleResultsInitializer = featureModuleResultsInitializer;
const moduleResultsInitializer = () => lodash.cloneDeep({
    ambiguous: 0,
    ambiguousPercentage: '',
    failed: 0,
    failedPercentage: '',
    passed: 0,
    passedPercentage: '',
    pending: 0,
    pendingPercentage: '',
    skipped: 0,
    skippedPercentage: '',
    status: status_1.Status.passed,
    time: '',
    timeInNanoSeconds: 0,
    total: 0,
    undefined: 0,
    undefinedPercentage: '',
});
exports.moduleResultsInitializer = moduleResultsInitializer;
const beforeOrAfterOverviewInitializer = () => lodash.cloneDeep({
    duration: 0,
    durationHHMMSS: '',
});
exports.beforeOrAfterOverviewInitializer = beforeOrAfterOverviewInitializer;
const overviewInitializer = () => lodash.cloneDeep({
    duration: 0,
    durationHHMMSS: '',
    status: status_1.Status.passed
});
exports.overviewInitializer = overviewInitializer;
const scenarioResultsInitializer = () => lodash.cloneDeep({
    after: (0, exports.beforeOrAfterOverviewInitializer)(),
    before: (0, exports.beforeOrAfterOverviewInitializer)(),
    overview: (0, exports.overviewInitializer)(),
    steps: (0, exports.moduleResultsInitializer)()
});
exports.scenarioResultsInitializer = scenarioResultsInitializer;
const featureResultsInitializer = () => lodash.cloneDeep({
    overview: {
        duration: 0,
        durationHHMMSS: '',
        result: [],
    },
    scenarios: (0, exports.moduleResultsInitializer)(),
    steps: (0, exports.moduleResultsInitializer)()
});
exports.featureResultsInitializer = featureResultsInitializer;
const reportResultsInitializer = () => lodash.cloneDeep({
    features: (0, exports.featureModuleResultsInitializer)(),
    overview: {
        date: new Date(),
        duration: 0,
        durationHHMMSS: '',
        metadata: [],
        metadataTitle: '',
        result: [],
        resultStatusesJoined: ''
    },
    scenarios: (0, exports.moduleResultsInitializer)(),
    steps: (0, exports.moduleResultsInitializer)()
});
exports.reportResultsInitializer = reportResultsInitializer;
