// cucumber-html-report-generator/dist/src/lib/helpers/console-messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.scenarioWithoutStepsRemoved = exports.scenarioWithoutIdRemoved = exports.featureRemoved = exports.invalidPathProvided = exports.invalidJsonProvided = exports.noJsonFilesProvided = exports.optionsNotProvided = exports.jsonPathNotProvided = exports.reportCreated = void 0;
const reportCreated = (featureId) => `\n 
    =====================================================================================
        Multiple Cucumber HTML report generated in:
        ${featureId}
\n
    =====================================================================================\n`;
exports.reportCreated = reportCreated;
exports.jsonPathNotProvided = 'A path which holds the JSON files should be provided.';
exports.optionsNotProvided = 'Options need to be provided.';
const noJsonFilesProvided = (jsonDir) => `No JSON files found in '${jsonDir}'. NO REPORT CAN BE CREATED!`;
exports.noJsonFilesProvided = noJsonFilesProvided;
const invalidJsonProvided = (fileNameAndPath, errorMessage) => `The file ${fileNameAndPath} is not a valid json, it has the error message ${errorMessage}`;
exports.invalidJsonProvided = invalidJsonProvided;
const invalidPathProvided = (filePath) => `The path provided: '${filePath}' is invalid`;
exports.invalidPathProvided = invalidPathProvided;
const featureRemoved = (featureName) => `The feature ${featureName} has been removed because it doesn't have any scenarios`;
exports.featureRemoved = featureRemoved;
const scenarioWithoutIdRemoved = (scenarioName, featureName) => `The scenario ${scenarioName} of the feature: ${featureName} has been removed as it doesn't have id.`;
exports.scenarioWithoutIdRemoved = scenarioWithoutIdRemoved;
const scenarioWithoutStepsRemoved = (scenarioName, featureName) => `The scenario ${scenarioName} of the feature: ${featureName} has been removed as it doesn't have steps.`;
exports.scenarioWithoutStepsRemoved = scenarioWithoutStepsRemoved;
