// cucumber-html-report-generator/dist/src/lib/generate-report/helpers/charts-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChartsData = void 0;
class ChartsData {
    constructor(suite) {
        this.dataFeatures = (suite) => ({
            datasets: [{
                    backgroundColor: [
                        '#228B22',
                        '#E74C3C',
                        '#FFD119',
                        '#3498DB',
                        '#b73122',
                        '#F39C12',
                        '#9b59b6'
                    ],
                    data: [
                        suite.results.features.passed,
                        suite.results.features.failed,
                        suite.results.features.pending,
                        suite.results.features.skipped,
                        suite.results.features.ambiguous,
                        suite.results.features.undefined,
                        suite.results.features.various
                    ],
                }],
            labels: [
                'Passed',
                'Failed',
                'Pending',
                'Skipped',
                'Ambiguous',
                'Undefined',
                'Various'
            ]
        });
        this.dataScenarios = (suite) => ({
            datasets: [{
                    backgroundColor: [
                        '#228B22',
                        '#E74C3C',
                        '#FFD119',
                        '#3498DB',
                        '#b73122',
                        '#F39C12'
                    ],
                    data: [
                        suite.results.scenarios.passed,
                        suite.results.scenarios.failed,
                        suite.results.scenarios.pending,
                        suite.results.scenarios.skipped,
                        suite.results.scenarios.ambiguous,
                        suite.results.scenarios.undefined,
                    ]
                }],
            labels: [
                'Passed',
                'Failed',
                'Pending',
                'Skipped',
                'Ambiguous',
                'Undefined'
            ]
        });
        this.dataSteps = (suiteOrFeature) => ({
            datasets: [{
                    backgroundColor: [
                        '#228B22',
                        '#E74C3C',
                        '#FFD119',
                        '#3498DB',
                        '#b73122',
                        '#F39C12'
                    ],
                    data: [
                        suiteOrFeature.results.steps.passed,
                        suiteOrFeature.results.steps.failed,
                        suiteOrFeature.results.steps.pending,
                        suiteOrFeature.results.steps.skipped,
                        suiteOrFeature.results.steps.ambiguous,
                        suiteOrFeature.results.steps.undefined
                    ]
                }],
            labels: [
                'Passed',
                'Failed',
                'Pending',
                'Skipped',
                'Ambiguous',
                'Undefined'
            ]
        });
        this.suite = suite;
    }
}
exports.ChartsData = ChartsData;
